docs(trainer): Add user guide for BuiltinTrainer(TorchTune LLM Trainer) (#4146)
kubeflow/website: content/en/docs/components/trainer/user-guides/builtin-trainer/torchtune.md

+++
title = "TorchTune BuiltinTrainer"
description = "How to fine-tune LLMs with TorchTune BuiltinTrainer"
weight = 20
+++

> This guide describes how to fine-tune LLMs using `BuiltinTrainer` and `TorchTuneConfig`. `TorchTuneConfig` leverages [TorchTune](https://github.com/pytorch/torchtune) to streamline LLMs fine-tuning on Kubernetes. To understand the concept of `BuiltinTrainer`, see the [overview guide](/docs/components/trainer/user-guides/builtin-trainer/overview.md)

Reviewer: ```suggestion
This guide describes how to fine-tune LLMs using `BuiltinTrainer` and `TorchTuneConfig`. `TorchTuneConfig` leverages [TorchTune](https://github.com/pytorch/torchtune) to streamline LLMs fine-tuning on Kubernetes. To understand the concept of `BuiltinTrainer`, see the [overview guide](/docs/components/trainer/user-guides/builtin-trainer/overview.md).
```

> The supported model list can be seen in [this directory](https://github.com/kubeflow/trainer/tree/master/manifests/base/runtimes/torchtune). It's worth noticing that we do not support multi-node fine-tuning for now. Only one Pod will be launched for TorchTune LLM Trainer.

Reviewer: Can you also use **Note** template, like you did for PVC ?

```suggestion
The supported model list can be seen in [this directory](https://github.com/kubeflow/trainer/tree/master/manifests/base/runtimes/torchtune). It's worth noticing that we do not support multi-node fine-tuning with TorchTune.
```

> In this page, we'll show a simple example of fine-tuning LLMs with TorchTune LLM Trainer in Kubeflow SDK. If you want to learn more about TorchTune LLM Trainer, please refer to [KEP-2401](https://github.com/kubeflow/trainer/tree/master/docs/proposals/2401-llm-trainer-v2) in Kubeflow Trainer.

Reviewer 2: ```suggestion
In this guide, we'll show a simple example of fine-tuning LLMs with TorchTune LLM Trainer in Kubeflow SDK. If you want to learn more about TorchTune LLM Trainer, please refer to [KEP-2401](https://github.com/kubeflow/trainer/tree/master/docs/proposals/2401-llm-trainer-v2) in Kubeflow Trainer.
```

## Prerequisites

Before exploring this guide, make sure to follow [the Getting Started guide](https://www.kubeflow.org/docs/components/trainer/getting-started/) to understand the basics of Kubeflow Trainer.

```Python
# Get torchtune-llama3.2-1b Kubeflow Training Runtime.
from kubeflow.trainer import *

client = TrainerClient()
runtime = client.get_runtime("torchtune-llama3.2-1b")
print(runtime)
```

Output:

```sh

> Runtime(name='torchtune-llama3.2-1b', trainer=Trainer(trainer_type=<TrainerType.BUILTIN_TRAINER: 'BuiltinTrainer'>, framework=<Framework.TORCHTUNE: 'torchtune'>, entrypoint=['tune', 'run'], accelerator='Unknown', accelerator_count='2'), pretrained_model=None)

Reviewer: Can you use updated output that shows two LLMs that we support w/ torchtune ?

```

> ## Use TorchTune LLM Trainer in BuiltinTrainer

Reviewer: As I asked here, can we not use LLM Trainer name ? We should change it to `TorchTuneConfig`: https://github.com/kubeflow/website/pull/4146#discussion_r2231476256
```suggestion
## Fine-Tune LLM with TorchTuneConfig
```

There are multiple places where you need to update it in this guide to be consistent.

Author: I'll update them to the consistent name. Thanks for the reminder:)

The guide below shows how to fine-tune Llama-3.2-1B-Instruct with alpaca dataset by BuiltinTrainer.

### Create PVCs for Models and Datasets

{{% alert title="Note" color="info" %}}

> Currently, we do not support automatically orchestrate the volume claim.

Reviewer 2: ```suggestion
Currently, we do not support automatically orchestrating the volume claim.
```

To follow up with this problem, please refer to [this issue](https://github.com/kubeflow/trainer/issues/2630).
{{% /alert %}}

We need to manually create PVCs for each models we want to fine-tune. Please note that **the PVC name must be equal to the ClusterTrainingRuntime name**. In this example, it's `torchtune-llama3.2-1b`.

```Python
# Create a PersistentVolumeClaim for the TorchTune Llama 3.2 1B model.
client.core_api.create_namespaced_persistent_volume_claim(
  body=client.V1PersistentVolumeClaim(
    api_version="v1",
    kind="PersistentVolumeClaim",
    metadata=client.V1ObjectMeta(name="torchtune-llama3.2-1b"),
    spec=client.V1PersistentVolumeClaimSpec(
      access_modes=["ReadWriteOnce"],
      resources=client.V1ResourceRequirements(
        requests={"storage": "20Gi"}
      ),
    ),
  ),
)
```

### Use TorchTune LLM Trainer with train() API

> You need to provide the following parameters to use TorchTune LLM Trainer with `train()` API:

Reviewer: Not all of the parameters are required.
I would suggest giving first simple example on how to fine-tune model without any changes:
```py
train(runtime=client.get_runtime("torchtune-llama3.2-1b")
```
And then expand the guide to explain how users can adjust the parameters for BuiltinTrainer.

Reviewer: Hmm, it might be tricky since we can't maintain the default args in the Controller.
Ok, for now let's ask users to configure data format and dataset in the `dataset_preprocess_config`.

- Reference to existing Runtimes.
- Initializer parameters.
- Trainer configuration.

> Kubeflow TrainJob will train the model in the referenced (Cluster)TrainingRuntime.

Reviewer: We don't support TrainingRuntime in SDK today, let's just call it: in the referenced TorchTune runtime, like we do for PyTorch: https://www.kubeflow.org/docs/components/trainer/user-guides/pytorch/#get-pytorch-runtime-packages

For example, you can use the `train()` API to fine-tune the Llama-3.2-1B-Instruct model using the alpaca dataset from HuggingFace with the code below:

```python
job_name = client.train(
    runtime=client.get_runtime("torchtune-llama3.2-1b"),
    initializer=Initializer(
        dataset=HuggingFaceDatasetInitializer(
            storage_uri="hf://tatsu-lab/alpaca/data"
        ),
        model=HuggingFaceModelInitializer(
            storage_uri="hf://meta-llama/Llama-3.2-1B-Instruct",
            access_token="<YOUR_HF_TOKEN>"  # Replace with your Hugging Face token
        )
    ),
    trainer=BuiltinTrainer(
        config=TorchTuneConfig(
            dataset_preprocess_config=TorchTuneInstructDataset(
                source=DataFormat.PARQUET,
            ),
            resources_per_node={
                "gpu": 1,
            }
        )
    )
)
```

### Watch the TrainJob Logs

> We can use the get_job_logs() API to get the TrainJob logs.

Reviewer: ```suggestion
We can use the `get_job_logs()` API to get the TrainJob logs.
```

#### Dataset Initializer

```python
from kubeflow.trainer.constants import constants

> log_dict = client.get_job_logs(job_name, follow=False, step=constants.DATASET_INITIALIZER)

Reviewer: This is not needed.

```suggestion
log_dict = client.get_job_logs(job_name, step=constants.DATASET_INITIALIZER)
```

> print(log_dict[constants.DATASET_INITIALIZER])

Reviewer: We might need to show sample output for the API.

```

#### Model Initializer

```python

> log_dict = client.get_job_logs(job_name, follow=False, step=constants.MODEL_INITIALIZER)

Reviewer: ```suggestion
log_dict = client.get_job_logs(job_name, step=constants.MODEL_INITIALIZER)
```

print(log_dict[constants.MODEL_INITIALIZER])
```

> #### TorchTune LLM Trainer

Reviewer: ```suggestion
#### Training Node
```

```python
log_dict = client.get_job_logs(job_name, follow=False)
print(log_dict[f"{constants.NODE}-0"])
```

### Get the Fine-Tuned Model

> After Trainer node completes the fine-tuning task, the fine-tuned model will be stored into the `/workspace/output ` directory, which can be shared across Pods through PVC mounting. You can find it in another Pod's `/<mountDir>/output` directory if you mount the PVC under `/<mountDir>`.

Reviewer: ```suggestion
After TrainJob completes the fine-tuning task, the fine-tuned model will be stored into the `/workspace/output ` directory, which can be shared across Pods through PVC mounting. You can find it in another Pod's `/<mountDir>/output` directory if you mount the PVC under `/<mountDir>`.
```

## Parameters

### Runtime

For TorchTune LLM Trainer, you can just find the runtime you want in the [manifest folder](https://github.com/kubeflow/trainer/tree/master/manifests/base/runtimes/torchtune), and just specify the `name` parameter. Like:

```python
runtime=Runtime(name="torchtune-llama3.2-1b")
```

### Initializer

We'll use parameters in Initializer to download dataset and model from remote storage.

| **Parameters** | **Type** | **What is it?** |
| - | - | - |
| `dataset` | `Optional[HuggingFaceDatasetInitializer]` | The configuration for one of the supported dataset initializers. |
| `model` | `Optional[HuggingFaceModelInitializer]` | The configuration for one of the supported model initializers. |

#### Dataset Initializer

Currently, we only support downloading datasets from HuggingFace by defining `HuggingFaceDatasetInitializer`:

{{% alert title="Note" color="info" %}}

> For `storage_uri` in Dataset Initializer, you need to speficy **the exact path to data files**. That's to say, you need to set it to `hf://<org_id>/<repo_id>/path/to/data/files`.

Reviewer 2: ```suggestion
For `storage_uri` in Dataset Initializer, you need to specify **the exact path to data files**. This means you need to set it to `hf://<org_id>/<repo_id>/path/to/data/files`.
```

Currently, we support:

> 1. Data Directory: Use all data files under this directory. For example, `hf://tatsu-lab/alpaca/data` uses all data files under the `/data` direcotry of `tatsu-lab/alpaca` repo in HuggingFace.

Reviewer 2: ```suggestion
1. Data Directory: Use all data files under this directory. For example, `hf://tatsu-lab/alpaca/data` uses all data files under the `/data` directory of `tatsu-lab/alpaca` repo in HuggingFace.
```

```Python
# Create a PersistentVolumeClaim for the TorchTune Llama 3.2 1B model.
client.core_api.create_namespaced_persistent_volume_claim(
  body=client.V1PersistentVolumeClaim(
    api_version="v1",
    kind="PersistentVolumeClaim",
    metadata=client.V1ObjectMeta(name="torchtune-llama3.2-1b"),
    spec=client.V1PersistentVolumeClaimSpec(
      access_modes=["ReadWriteOnce"],
      resources=client.V1ResourceRequirements(
        requests={"storage": "20Gi"}
      ),
    ),
  ),
)
```

### Use TorchTune BuiltinTrainer with train() API

You need to provide the following parameters to use TorchTune BuiltinTrainer with `train()` API:

- Reference to existing Runtimes.
- Initializer parameters.
- Trainer configuration.

Kubeflow TrainJob will train the model using the referenced TorchTune runtime `torchtune-llama3.2-1b`.

For example, you can use the `train()` API to fine-tune the Llama-3.2-1B-Instruct model using the alpaca dataset from HuggingFace with the code below (need 2 GPUs):

```python
job_name = client.train(
    runtime=client.get_runtime("torchtune-llama3.2-1b"),
    initializer=Initializer(
        model=HuggingFaceModelInitializer(

>             storage_uri="hf://meta-llama/Llama-3.2-1B-Instruct",

Author: It's a must to have parameter, since we'll do the override here anyway:

https://github.com/kubeflow/sdk/blob/7614d748d3046750a5ae12a55ddac10db9fd8e7d/python/kubeflow/trainer/utils/utils.py#L574-L579

Maybe we can update the logic here? WDYT? @andreyvelich

+++
title = "TorchTune LLM Trainer"
description = "How to use TorchTune LLM Trainer in BuiltinTrainer"
weight = 20
+++

> **TorchTuneConfig**, configurations for **TorchTune LLM Trainer**, is one of the supported configs in `BuiltinTrainer`. It leverages TorchTune to streamline and simplify LLM fine-tuning lifecycle on Kubernetes, while providing a simple yet efficient Python API for data scientists.

Reviewer 3: ```suggestion
**TorchTuneConfig**, the configuration for **TorchTune LLM Trainer**, is one of the supported configs in `BuiltinTrainer`. It leverages TorchTune to streamline and simplify LLM fine-tuning lifecycle on Kubernetes, while providing a simple yet efficient Python API for data scientists.
```

> **TorchTuneConfig**, the configuration for **TorchTune LLM Trainer**, is one of the supported configs in `BuiltinTrainer`. It leverages TorchTune to streamline and simplify LLM fine-tuning lifecycle on Kubernetes, while providing a simple yet efficient Python API for data scientists.

Author: Do you mean that we should define `TorchTuneConfig` as "the configuration for TorchTune BuiltinTrainer"?

> **TorchTuneConfig**, the configuration for **TorchTune LLM Trainer**, is one of the supported configs in `BuiltinTrainer`. It leverages TorchTune to streamline and simplify LLM fine-tuning lifecycle on Kubernetes, while providing a simple yet efficient Python API for data scientists.

Reviewer: Yes, let's be consistent with naming and say:
```
This guide describes how to fine-tune LLMs using `BuiltinTrainer` and `TorchTuneConfig`.
The `TorchTuneConfig` leverages [TorchTune](https://github.com/pytorch/torchtune) to streamline
LLMs fine-tuning lifecycle on Kubernetes.

To understand the concept of `BuiltinTrainer`, see the [overview guide](...)
```

The supported model list can be seen in [this directory](https://github.com/kubeflow/trainer/tree/master/manifests/base/runtimes/torchtune). It's worth noticing that we do not support multi-node fine-tuning for now. Only one Pod will be launched for TorchTune LLM Trainer.

In this page, we'll show a simple example of fine-tuning LLMs with TorchTune LLM Trainer in Kubeflow SDK. If you want to learn more about TorchTune LLM Trainer, please refer to [KEP-2401](https://github.com/kubeflow/trainer/tree/master/docs/proposals/2401-llm-trainer-v2) in Kubeflow Trainer.

## Prerequisites

Before exploring this guide, make sure to follow [the Getting Started guide](https://www.kubeflow.org/docs/components/trainer/getting-started/) to understand the basics of Kubeflow Trainer.

```Python
# Get torchtune-llama3.2-1b Kubeflow Training Runtime.
from kubeflow.trainer import *

client = TrainerClient()
runtime = client.get_runtime("torchtune-llama3.2-1b")

> print(runtime)

Author: Sure. I'll add it

```

## Use TorchTune LLM Trainer in BuiltinTrainer

> The guides below shows how to fine-tune Llama-3.2-1B-Instruct with alpaca datset by BuiltinTrainer.

Reviewer 2: ```suggestion
The guide below shows how to fine-tune Llama-3.2-1B-Instruct with alpaca dataset by BuiltinTrainer.
```

2. Single Data File: Use the single data file given the path. For example, `hf://tatsu-lab/alpaca/data/xxx.parquet` uses the single `/data/xxx.parquet` data file of `tatsu-lab/alpaca` repo in HuggingFace.
{{% /alert %}}

| **Parameters** | **Type** | **What is it?** |
| - | - | - |
| `storage_uri` | `str` | Storage URI for dataset that begins with `hf://` |
| `access_token` | `Optional[str]` | Token that verfies your access to the dataset if needed. |

#### Model Initializer

Currently, we only support downloading models from HuggingFace by defining `HuggingFaceModelInitializer`:

| **Parameters** | **Type** | **What is it?** |
| - | - | - |
| `storage_uri` | `str` | Storage URI for model that begins with `hf://` |
| `access_token` | `Optional[str]` | Token that verfies your access to the model if needed. |

#### Example Usage

```python
initializer=Initializer(
    dataset=HuggingFaceDatasetInitializer(
        storage_uri="hf://tatsu-lab/alpaca/data",
        access_token="<YOUR_HF_TOKEN>"  # Replace with your Hugging Face token
    ),
    model=HuggingFaceModelInitializer(
        storage_uri="hf://meta-llama/Llama-3.2-1B-Instruct",

>         access_token="<YOUR_HF_TOKEN>"  # Replace with your Hugging Face token

Reviewer 2: Or maybe we should add this to the parameter table?

>         access_token="<YOUR_HF_TOKEN>"  # Replace with your Hugging Face token

Author: > To get token go to [Hugging Face](https://huggingface.co/settings/tokens) and click on "Create new token".

From my perspective, it might be too detailed, because I assume that every data scientists know HF and HF token. But I'm glad to add it to the parameter table if you consider it necessary:)

>         access_token="<YOUR_HF_TOKEN>"  # Replace with your Hugging Face token

Reviewer 2: I think you're right :)

    )
)
```

### TorchTune LLM Trainer

#### Description

> The `TorchTuneConifg` class is used for configuring TorchTune LLM Trainer that already includes the fine-tuning logic.

Reviewer 3: ```suggestion
The `TorchTuneConfig` class is used for configuring TorchTune LLM Trainer that already includes the fine-tuning logic.
```

| **Parameters** | **Type** | **What is it?** |
| - | - | - |
| `dtype` | `Optional[DataType]` | The underlying data type used to represent the model and optimizer parameters. Currently, we only support `bf16` and `fp32`. |

> | `batch_size` | `Optional[int]` | The number of samples processed before updating model weights. |
> | `epochs` | `Optional[int]` | The number of samples processed before updating model weights. |

Reviewer 2: ```suggestion
| `batch_size` | `Optional[int]` | The number of samples processed before updating model weights. |
| `epochs` | `Optional[int]` | The number of times the entire dataset is processed during training. |
```

| `loss` | `Optional[Loss]` | The loss algorithm we use to fine-tune the LLM, e.g. `torchtune.modules.loss.CEWithChunkedOutputLoss` |
| `num_nodes` | `Optional[int]` | The number of PyTorch Nodes in training |
| `dataset_preprocess_config` | `Optional[TorchTuneInstructDataset]` | Configuration for dataset preprocessing. |

> | `resource_per_node` | `Optional[Dict]` | The resource for each PyTorch Node |

Author: @kramaranya Nice catch!

> | `resources_per_node` | `Optional[Dict]` | The resource for each PyTorch Node |

Reviewer: From my past experience I found it hard to keep such doc up-to-date with SDK docstrings.
My suggestions is to keep SDK docstring updated for the TorchTuneConfig, and redirect users to the code base (at least for now), rather than maintain the table in the website

+++
title = "TorchTune LLM Trainer"
description = "How to use TorchTune LLM Trainer in BuiltinTrainer"
weight = 20
+++

> **TorchTune LLM Trainer** is one of the supported BuiltinTrainers in Kubeflow SDK. It leverages TorchTune to streamline and simplify LLM fine-tuning lifecycle on Kubernetes, while providing a simple yet efficient Python API for data scientists.

Author: I'll rephrase the expressions according to your advice and my concern here: https://github.com/kubeflow/website/pull/4146#discussion_r2224255367

```bash
pip install git+https://github.com/kubeflow/sdk.git@main#subdirectory=python
```

## Use TorchTune LLM Trainer in BuiltinTrainer

### Create PVCs for Models and Datasets

{{% alert title="Note" color="info" %}}
Currently, we do not support automatically orchestrate the volume claim.

To follow up with this problem, please refer to [this issue](https://github.com/kubeflow/trainer/issues/2630).
{{% /alert %}}

We need to manually create PVCs for each models we want to fine-tune. Please note that **the PVC name must be equal to the ClusterTrainingRuntime name**. In this example, it's `torchtune-llama3.2-1b`.

```Python
# Create a PersistentVolumeClaim for the TorchTune Llama 3.2 1B model.
client.core_api.create_namespaced_persistent_volume_claim(
  namespace="default",
  body=client.V1PersistentVolumeClaim(
    api_version="v1",
    kind="PersistentVolumeClaim",
    metadata=client.V1ObjectMeta(name="torchtune-llama3.2-1b"),
    spec=client.V1PersistentVolumeClaimSpec(
      access_modes=["ReadWriteOnce"],
      resources=client.V1ResourceRequirements(
        requests={"storage": "20Gi"}
      ),
    ),
  ),
)
```

> ### Use TorchTune LLM Trainer with `train()` API

Reviewer: Let's not use the `<code>` block in the titles, since the website renders it blurry
